modules/core/src/main/java/org/apache/ignite/cache/affinity/fair/FairAffinityFunction.java

```diff
@@ -865,16 +896,17 @@ else if (force) {
          * @param tier Tier.
          * @param node Node.
          * @param allowNeighbors Allow neighbors.
+         * @return Can partition assign to node.
          */
         private boolean isAssignable(int part, int tier, final ClusterNode node, boolean allowNeighbors) {
             if (containsPartition(part, node))
                 return false;
 
             if (exclNeighbors)
                 return allowNeighbors || !neighborsContainPartition(node, part);
-            else if (backupFilter == null)
-                return true;
-            else {
+            else if (affinityBackupFilter != null)
+                return affinityBackupFilter.apply(node, assignments.get(part));
```

Comment:
This implementation is incorrect because you need to consider "tier" parameter passed to the function.

 Before affinityBackupFilter.apply() is executed you should remove an already probably assigned value from assignments.get(part) and pass it to the filter. See how it's done from backup filter and tier=0 (primary node). 

Also due to this fact list of the nodes that is passed to affinityBackupFilter in this affinity function may not contain primary as the first element in the list.